implement the database setup task

lurldgbodex opened the pull request:
## How to test this PR

### Cloning the repository
```bash
git clone https://github.com/hngprojects/hng_boilerplate_java_web.git
cd hng_boilerplate_java_web
git checkout chore/database-setup
```

### Setting Up the Environment

#### Prerequisites:

   - Ensure you have Java 17 installed.
   - Install Maven (or Gradle, depending on your project).

#### Database Setup
- Ensure PostgreSQL is installed and running.
- Create a new database
  `CREATE DATABASE {database_name}`

#### Update the application.properties file with your database credentials
```
spring.datasource.url=jdbc:postgresql://localhost:5432/your_database_name
spring.datasource.username=your_username
spring.datasource.password=your_password
```
### Running the Application

#### Build and run the application using Maven
```bash
mvn clean install
mvn spring-boot:run
```
### Flyway Migration
#### Configure Flyway
- update the application.properties file to configure flyway
```html
spring.flyway.enabled=true
spring.flyway.locations=classpath:db/migration
spring.flyway.user={database_username}
spring.flyway.password={database_password}
```
#### Create Migration Files
- Migration files should be placed in the `src/main/resources/db/migration` directory. 
- Each migration file should follow the naming convention `V<version>__<description>.sql`. e.g.
  -  V1__init_database.sql

#### Sample Migration file
```sql
CREATE TABLE users (
    id VARCHAR(255) PRIMARY KEY,
    name VARCHAR(255) NOT NULL,
    email VARCHAR(255) NOT NULL UNIQUE
);

-- V2__create_profiles_table.sql
CREATE TABLE profiles (
    id VARCHAR(255) PRIMARY KEY,
    first_name VARCHAR(255),
    last_name VARCHAR(255),
    phone VARCHAR(255),
    avatar_url VARCHAR(255)
);

-- V3__create_products_table.sql
CREATE TABLE products (
    id VARCHAR(255) PRIMARY KEY,
    name VARCHAR(255) NOT NULL,
    description TEXT,
    user_id VARCHAR(255) NOT NULL,
    FOREIGN KEY (user_id) REFERENCES users(id)
);

-- V4__create_organisations_table.sql
CREATE TABLE organisations (
    id VARCHAR(255) PRIMARY KEY,
    name VARCHAR(255) NOT NULL,
    description TEXT
);

-- Many-to-Many Relationship Table
CREATE TABLE user_organisation (
    user_id VARCHAR(255) NOT NULL,
    organisation_id VARCHAR(255) NOT NULL,
    PRIMARY KEY (user_id, organisation_id),
    FOREIGN KEY (user_id) REFERENCES users(id),
    FOREIGN KEY (organisation_id) REFERENCES organisations(id)
);
```
#### Apply Migrations
- Flyway will automatically apply the migrations on application startup. 
- Ensure that the application starts without errors and that the tables are created as expected.

##### To manually apply migrations:
`./mvnw flyway:migrate`

## Testing the Endpoint

### You can test the endpoint using curl or Postman
- Using Curl (user_one_id: the id of the user created in the database. You can get the id by querying the database for you created user)
`curl -X GET http://localhost:8080/api/v1/users/user_one_id`

- Using Postman
  * Open Postman.
  * Create a new GET request to http://localhost:8080/api/v1/users/user_one_id.
  * Send the request and verify the response.
- Expected Response
```json
{
  "name": "John Doe",
  "id": "some-user-id",
  "email": "johndoe@example.com",
  "profile": {
    "first_name": "John",
    "last_name": "Doe",
    "phone": "1234567890",
    "avatar_url": "http://example.com/avatar.jpg"
  },
  "organisations": [
    {
      "org_id": "some-org-id-1",
      "name": "Some Org",
      "description": "Some Org Description"
    },
    {
      "org_id": "some-other-org-id-2",
      "name": "Some Other Org",
      "description": "Some Other Org Description"
    }
  ],
  "products": [
    {
      "product_id": "some-product-id",
      "name": "Some Product",
      "description": "Some Product Description"
    },
    {
      "product_id": "some-other-product-id",
      "name": "Some Other Product",
      "description": "Some Other Product Description"
    }
  ]
}
```

- There are 2 default users created when you start the application. 
- To get the users, login to your database and query the users table using:
`SELECT * FROM users;`
- copy the user_id and make the get request

nedssoft commented:
LGTM

NdubuisiJr commented on src/main/resources/application.properties:
```diff
@@ -1 +1,7 @@
 spring.application.name=hng-java-boilerplate
+
+# datasource configuration
+spring.datasource.url=jdbc:postgresql://localhost/{database}
```
This whole url should be an environment variable. We don't need `localhost` here.

lurldgbodex replied:
Okay. I will effect the changes now

NdubuisiJr commented:
@lurldgbodex Please update the `READMe` file to instruct other interns on how to setup their database locally. Something similar to what you have on this PR.

NdubuisiJr commented:
@lurldgbodex you haven't resolved my review comment.

lurldgbodex commented:
I updated the readme under setup instruction  and added the setup database, and i also updated the verify setup to verify the database setup. It's in the second commit from line 163

NdubuisiJr commented:
@lurldgbodex click this [link](https://github.com/hngprojects/hng_boilerplate_java_web/pull/13/files#r1683421091). Check the comment on line 4 in `application.properties`

lurldgbodex commented:
@NdubuisiJr The change has been made

NdubuisiJr commented:
@lurldgbodex good work so far! 
You have two bugs to fix
1. Each time the app starts data is `seeded` into the db. That's wrong, it should only happen when the db is empty.
2. Each time a model (entity) is changed, the database schema is automatically updated without a migration. This is wrong, schemas should only be modified by using a migration.
<img width="1024" alt="Screenshot 2024-07-18 at 23 34 30" src="https://github.com/user-attachments/assets/e3c12351-ea9a-49db-8439-6564ae1ee06f">

You can use setup flyway for database migrations and disable auto schema update.

lurldgbodex commented:
@NdubuisiJr  I have implement flyway migration and updated the seeding of database to only happen when the database is empty
